sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -620,6 +620,33 @@ object PushFoldableIntoBranches extends Rule[LogicalPlan] with PredicateHelper {
   }
 }
 
+/**
+ * Remove duplicated case when branches.
+ */
+object RemoveDuplicatedBranches extends Rule[LogicalPlan] with PredicateHelper {
+
+  /**
+   * Wrapper around a branch that provides semantic equality.
+   */
+  case class EquivalentBranch(br: (Expression, Expression)) {
+    override def equals(o: Any): Boolean = o match {
+      case other: EquivalentBranch if br._1.deterministic && other.br._1.deterministic =>
+        br._1.semanticEquals(other.br._1)
+      case _ => false
+    }
+
+    override def hashCode: Int = br._1.semanticHash()
+  }
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case q: LogicalPlan => q transformExpressionsUp {
+      case c @ CaseWhen(branches, _) if branches.length > 1 =>
+        val equivalentBranchSet = branches.map(EquivalentBranch).toSet
+        val dedup = equivalentBranchSet.map(_.br).toSeq
```

Comment:
Is it perhaps more performant to not copy if there were no duplicates? not sure.

Reply:
Checking no duplicates like this needs to use `Set.size`, but this will traverse the set.
```
        if (equivalentBranchSet.size < branches.length) {
          val dedup = equivalentBranchSet.map(_.br).toSeq
          c.copy(branches = dedup)
        } else {
          c
        }
```

Reply:
Sure, but isn't that cheaper than .toSeq every time?

Reply:
Fixed.